### ISSUE TITLE
[Edit Mode Prototype] Create button initiates edit-mode immediately

### ISSUE THREAD

**Issue by akhenry** [nasa/openmctweb#286]
From https://github.com/nasa/openmctweb/issues/161#issuecomment-155162928 No. 10
User is not taken through create dialog before being allowed to edit. Instead, create dialog is shown on save

**Comment by akhenry** [comment]
@VWoeltjen In implementing this, I'm adding a dependency from the CreateAction to Instantiate so that the action can instantiate a new in-memory object for editing. Previously a new persisted object would have been created on completion of the create wizard, using the creation capability of the parent object.

Adding a dependency on Instantiate has created a circular dependency. The chain is something like - CreateAction -> Instantiate -> CapabilityService -> Capabilities[] -> ActionCapability -> ActionService -> ActionAggregator -> Actions[] -> CreateAction.

Circular dependencies are typical an opportunity for improvements in architecture :) The key link in the chain to me is the implicit dependency from capabilities to actions. This feels a little wrong. The two should not be dependent in both directions.

A dependency on capabilities from actions is ok - actions utilize an object's capabilities. But it doesn't feel right conceptually to have a dependency in the other direction. Exposing an object's actions via a capability is a convenience, but a really useful one, so we should keep it. I wonder however if perhaps the ActionCapability could lazy-load the actionService instead?

A second option would be to have Instantiate lazy-load the capabilities service, but I'm not sure that makes sense. Object instantiation conceptually requires access to capabilities.

**Comment by VWoeltjen** [comment]
Some additional options:

* The "magic bullet" for circular dependencies is `$injector`; CreateAction could depend on it and lazily initialize your dependencies.
* CreateAction could also use the `instantiation` capability of the parent object instead of the `instantiate` service. This is a little preferable, actually, since it means that the new object will be instantiated with an appropriate persistence space encoded in its identifier, without having to put any particular thought into doing so from CreateAction.  (The `instantiation` capability itself uses the `$injector` approach.)
  * May be useful to add a note to the JSDoc of `instantiate` indicating that usage the `instantiation` capability is preferred, and recommending using `instantiate` directly only when there is no appropriate object from which to retrieve that capability.
* Could also move the `$injector` trick from the `instantiation` capability to the `instantiate` service (which would break the dependency here), not sure that this makes any more sense.

Actions are most commonly upon domain object instances; as you stated, there is a logical dependency between instantiation of domain objects and the `capabilityService`, but once you have an instance, you generally just use the capabilities.

I see your point about the action capability being more of a "convenience" than an intrinsic dependency, but it's also consistent with a general pattern among capabilities. Very commonly, capabilities are your object-specific "window" onto the service layer - `domainObject.getCapability('persistence')` gives you an interface for making calls to the `persistenceService` without worrying about which specific details of `domainObject` should be passed in, `domainObject.getCapability('telemetry')` does similar relative to `telemetryService`, and so on. In this regard, it seems valid for the `capabilityService` to depend on any/all services which have behavior applicable to domain objects. (Of course, this extra layer of abstraction introduces confusion along with convenience, and is due for some significant re-thinking, but I'd consider that a concern for #69.)

In other words, I don't see any inherent reason why `actionService` should depend on `capabilityService`, but do see at least a weak argument for `capabilityService` to depend on `actionService`. Given that, I'd be inclined to break the dependency somewhere between `actionService` and `capabilityService`, either by using the `instantiation` capability in CreateAction, or loading `instantiate` via `$injector`.

**Comment by akhenry** [comment]
Thanks for the feedback. I think the easiest option here is to use the instantiation capability of the object currently selected in the tree. This makes the assumption however that there will _always_ be something selected in the navigation tree on the left. Is this a safe assumption to make?

**Comment by larkin** [comment]
I think we should get rid of the CreateAction.  It's not something you perform on a domain object-- it creates a domain object.  Previously, actions have always been performed on domain objects.

Again-- this is where I think we are overloading terms.  I've suggested before "ContextMenuActions" separate from "Actions", perhaps we need to clarify "DomainObjectActions" separate from "Actions"?  

This type registration is really getting in our way.  Seeing as there is very little commonality between actions, it seems like it's causing more issues than it is solving.

I'd also absolutely prefer lazyloading of capabilities.

And, to add to the shopping cart, I'd also prefer that we throw out `Instantiate` and toss it's behavior in the constructor where it belongs. I suppose it's worth highlighting that the angular injector is forcing us to rewrite basic javascript constructs(`new`) because a singleton that should be defined with requirejs (the capabilityService) is provided by angular and we've got two different dependency management/injection systems which forces us to to write an entire service called `instantiate` just because we can't otherwise reference the `capabilityService` at runtime.

Baseline, think this is a lot of extraneous code which is intended to support reusability when it's not actually reusable.

**Comment by VWoeltjen** [comment]
Moving some of this over to #69 since we're stepping a little bit out of scope for this; we're having multiple good discussions but they're separate from "Create button initiates new edit mode immediately"

**Comment by akhenry** [comment]
A prototype of this functionality has been implemented in [branch Open199](https://github.com/nasa/openmctweb/tree/open199)

**Comment by akhenry** [comment]
Closing this issue. Conversation about API is continued on #69
